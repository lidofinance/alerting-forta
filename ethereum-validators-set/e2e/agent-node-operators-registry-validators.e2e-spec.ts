import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-node-operators-registry e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (
    txHash: string,
    initBlock?: number,
  ) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;
  let timeSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
    timeSpy = jest.spyOn(Date, "now");
    timeSpy.mockImplementation(() => new Date("2023-12-31"));
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath(
          "subagents/node-operators-registry/agent-node-operators-registry",
        ),
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process tx with removed signing keys",
    async () => {
      const findings = await runTransaction(
        "0x43529ea1c44b6f6134fa045f6af81a1a15be6309a32424ea742fa056977c4010",
      );
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with csm removed signing keys",
    async () => {
      const findings = await runTransaction(
        "0x2dfe92fad0397179082990bce51633d8334aa729d4566f6392a988c7fce7c2be",
      );
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with a lot of validators exited",
    async () => {
      const findings = await runTransaction(
        "0xdc70082a674abb83a94f04ea5b083849ff476ddc2f0fdd2a1cd2f7d2e079592b",
        19454500,
      );
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );
});
