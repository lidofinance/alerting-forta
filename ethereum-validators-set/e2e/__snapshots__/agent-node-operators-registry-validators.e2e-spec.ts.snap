// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-node-operators-registry e2e tests should process tx with a lot of validators exited 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #13 Blockdaemon
New exited: 117",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #20 HashKey Cloud
New exited: 167",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #23 CryptoManufaktur
New exited: 117",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #24 Kukis Global
New exited: 117",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #25 Nethermind
New exited: 116",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #26 ChainSafe
New exited: 216",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #27 Prysm Team at Offchain Labs
New exited: 116",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #28 Sigma Prime
New exited: 116",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-BIG-EXIT",
    "description": "Operator: #29 Attestant (BVI) Limited
New exited: 116",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: operator exited more than 100 validators",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`agent-node-operators-registry e2e tests should process tx with csm removed signing keys 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "COMMUNITY-STAKING-NODE-OPERATORS-KEYS-REMOVED",
    "description": "Signing keys has been removed for:
 CSM Operator #333: 25",
    "labels": [],
    "metadata": {},
    "name": "⚠️ Community Staking: Signing keys removed",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`agent-node-operators-registry e2e tests should process tx with removed signing keys 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "CURATED-ONCHAIN-V1-NODE-OPERATORS-KEYS-REMOVED",
    "description": "Signing keys has been removed for:
 Kiln: 50",
    "labels": [],
    "metadata": {},
    "name": "⚠️ curated-onchain-v1: Signing keys removed",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2023-12-31T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;
